### PR TITLE
Refactor `PetscVector` to use buffered writes

### PR DESCRIPTION
**Pull request by ZedThree**
In recent versions of PETSc, we get tens of thousands of error messages due to mixing `VecGetValue` and `VecSetValue` calls -- although everything does still seem to work fine.

This PR replaces the `PetscVector::Element` proxy class with an internal buffer. Setting vector elements fills up the buffer, and then calls to `PetscVector::assemble` passes the whole buffer to PETSc at once.

Unfortunately, this results in a degrading of the result for reasons I've not been able to work out. Everything we pass to PETSc has remained identical, but the solution changes by 1e-11 -- and after applying the forward operator to check the result, this blows up to a difference of 1e-2, enough to put it over our test's tolerance.

I need to move on to other things, but if anyone has any suggestions, that would be great!

I've also given some general tiding to `LaplacePetsc3Damg`, because it was bugging me.

### CODE REVIEW

**Comment by ZedThree** [comment]
I think most of the diff is from changed output files that they keep in their tests.

I've managed to dump the `rhs` and `guess` vectors, and the `operator3D` matrix for this branch and current `master`. The vectors have identical values, but there are differences in the matrices. The first difference is at row 1024 (I've stripped out duplicate columns, so this is just the differences with some context):

```diff
$ diff <(grep "row 1024:" $oldmatrix ) <(grep "row 1024:" matrix.dat )
1c1
< row 1024: ...  (993, -0.)  (1022, -0.)  (1023, 0.)  ...  (1087, 15.5001)  (2016, 30.6409)  (2017, 15.5041)  (2018, -0.0714378)  (2045, 0.000670876)  (2046, -0.0767834)  (2047, -15.4969)  (2048, 1.15936e+07) ... (3041, 0.)  (3070, 0.)  (3071, 0.) 
---
> row 1024: ...  (993, 0.)  (994, -0.)  (1023, -0.)  ...  (1087, 15.5001)  (2016, 30.3472)  (2017, 15.791)  (2018, 0.0740949)  (2019, -0.000669586)  (2046, 0.0713701)  (2047, -15.7825)  (2048, 1.15936e+07) ... (3041, 0.)  (3042, 0.)  (3071, 0.)
$ diff <(grep "row 1025:" $oldmatrix ) <(grep "row 1025:" matrix.dat )
1c1
< row 1025: ... (992, 0.)  (993, 0.)  (994, -0.)  (1023, -0.)  (1024, 32.3711) ... (2016, -15.5075)  (2017, 30.4052)  (2018, 15.5147)  (2019, -0.0714813)  (2046, 0.000671284)  (2047, -0.0768193)  (2048, -0.)  ...  (3042, 0.)  (3071, 0.) 
---
> row 1025: ... (992, -0.)  (993, 0.)  (994, 0.)  (995, -0.)  (1024, 32.3711) ... (2016, -15.7909)  (2017, 30.1113)  (2018, 15.7994)  (2019, 0.0741509)  (2020, -0.000669994)  (2047, 0.0714136)  (2048, -0.)  ...  (3042, 0.)  (3043, 0.) 
$ diff <(grep "row 1026:" $oldmatrix ) <(grep "row 1026:" matrix.dat )
1c1
< row 1026: ... (36, 0.)  (992, -0.)  (993, 0.)  (994, 0.)  (995, -0.)  (1025, 32.2787) ... (1061, 0.000674656)  (2016, -0.0773674)  (2017, -15.6139)  (2018, 30.9309)  (2019, 15.6212)  (2020, -0.0719788)  (2047, 0.000675956)  (2049, -0.)  (2050, 1.16981e+07) ... 
---
> row 1026: ... (36, 0.)  (993, -0.)  (994, 0.)  (995, 0.)  (996, -0.)  (1025, 32.2787) ... (1061, 0.000674656)  (2016, 0.0719106)  (2017, -15.9022)  (2018, 30.635)  (2019, 15.9108)  (2020, 0.0746535)  (2021, -0.000674656)  (2049, -0.)  (2050, 1.16981e+07) ...
```

Annoyingly, PETSc doesn't dump matrices to hdf5, so I'm having to use text.

Definitely getting something wrong here! Some column indices are different, off by 26/28/29, and some values are wrong. So I've probably messed up the indexing somehow.

**Comment by ZedThree** [comment]
I think I've got this working now! It's a silly error on my part, I generalised
`ShiftedMetricInterp::getWeightsForYApproximation` but didn't update the
parallel slice for a general `y_offset`

**Comment by dschwoerer** [comment]
The failures are only partially due to fedora. For some reason the fedora CI failure does not fail the other ones.

The real failure is:
```
 In file included from /home/runner/work/BOUT-dev/BOUT-dev/src/invert/laplace/impls/petsc3damg/petsc3damg.hxx:50,
                 from /home/runner/work/BOUT-dev/BOUT-dev/src/invert/laplace/impls/petsc3damg/petsc3damg.cxx:31:
/home/runner/work/BOUT-dev/BOUT-dev/include/bout/petsc_interface.hxx: In instantiation of ‘PetscMatrix<T>::Element PetscMatrix<T>::operator()(const ind_type&, const ind_type&) [with T = Field3D; PetscMatrix<T>::ind_type = SpecificInd<IND_TYPE::IND_3D>]’:
/home/runner/work/BOUT-dev/BOUT-dev/src/invert/laplace/impls/petsc3damg/petsc3damg.cxx:126:20:   required from here
/home/runner/work/BOUT-dev/BOUT-dev/include/bout/petsc_interface.hxx:448:70: error: use of deleted function ‘PetscMatrix<T>::Element::Element(PetscMatrix<T>::Element&&) [with T = Field3D]’
  448 |     return Element(matrix.get(), global1, global2, positions, weights);
      |                                                                      ^
/home/runner/work/BOUT-dev/BOUT-dev/include/bout/petsc_interface.hxx:339:5: note: declared here
  339 |     Element(Element&&) = delete;
      |     ^~~~~~~
```
which seems related to this PR ...

**Comment by ZedThree** [comment]
Ah, thanks, didn't spot that! That's a bit odd. It looks like most of our runners are still on ubuntu-20.04 for some reason, which uses gcc 9.4, so looks like it might be a compiler issue? I think I deleted the move ctor on purpose, but I'll double check

**Comment by ZedThree** [comment]
Oh, probably newer compilers are just eliding the move completely. Hopefully just `=default`-ing the move ctors is sufficient

File: include/bout/petsc_interface.hxx [hunk 1]
```diff
@@ -441,7 +400,7 @@ public:
       std::transform(weights.begin(), weights.end(), std::back_inserter(values),
                      [&val](BoutReal weight) -> PetscScalar { return weight * val; });
 
-      int status;
+      int status = 0;
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply include/bout/petsc_interface.hxx:403]: Clang-tidy, you are very confused today

**Comment by ZedThree** [comment]
[694c245](https://github.com/boutproject/BOUT-dev/pull/2713/commits/694c245be5474fc65448369ed8896092d79ca218) also fixes #2755

**Comment by ZedThree** [comment]
Woop, that failing PETSc CI build is finally working!

**Review by dschwoerer — APPROVED** [review]
Awesome :+1: 
Looks much cleaner nice, thank you also for tidying up :-)

**Comment by dschwoerer** [comment]
As that is a bug fix, should it also go into master?
Or should we disable recent PETSc  in master?

**Comment by ZedThree** [comment]
I think just disable recent PETSc in master for now

**Comment by dschwoerer** [comment]
In local testing https://github.com/boutproject/BOUT-dev/pull/2781 did resolve the test failure.
That should be small enough to go into master (finger crossed)

**Comment by tomc271** [comment]
Fixes #2705

**Pull request merged**